### ISSUE TITLE
Issues with recovering from closed Pika connections

### ISSUE THREAD

**Issue by whalesalad** [Bogdanp/dramatiq#36]
Here is the traceback:

```
ERROR:pika.adapters.base_connection:Read empty data, calling disconnect
ERROR:pika.adapters.blocking_connection:Connection close detected; result=BlockingConnection__OnClosedArgs(connection=<SelectConnection CLOSED socket=None params=<URLParameters host=rabbitmq.farmlogs port=5672 virtual_host=planet ssl=False>>, reason_code=-1, reason_text='EOF')
Traceback (most recent call last):
  File "bin/cron", line 79, in <module>
    schedule.run_pending()
  File "/usr/local/lib/python3.6/dist-packages/schedule/__init__.py", line 452, in run_pending
    default_scheduler.run_pending()
  File "/usr/local/lib/python3.6/dist-packages/schedule/__init__.py", line 75, in run_pending
    self._run_job(job)
  File "/usr/local/lib/python3.6/dist-packages/schedule/__init__.py", line 129, in _run_job
    ret = job.run()
  File "/usr/local/lib/python3.6/dist-packages/schedule/__init__.py", line 377, in run
    ret = self.job_func()
  File "bin/cron", line 63, in check_and_process_new_flags
    process_enterprise.send(enterprise_id)
  File "/usr/local/lib/python3.6/dist-packages/dramatiq/actor.py", line 104, in send
    return self.send_with_options(args=args, kwargs=kwargs)
  File "/usr/local/lib/python3.6/dist-packages/dramatiq/actor.py", line 129, in send_with_options
    return self.broker.enqueue(message, delay=delay)
  File "/usr/local/lib/python3.6/dist-packages/dramatiq/brokers/rabbitmq.py", line 206, in enqueue
    raise ConnectionClosed(e) from None
dramatiq.errors.ConnectionClosed: (-1, 'EOF') 
```

I'm wondering if this has something to do with the ELB that I have in front of my RabbitMQ hosts. These connection failures happen all the time and then work gets dropped on the floor. I'm not sure if this is an issue I need to try and patch at the Pika level or the Dramatiq level.

Here is a thread from the Pika issue tracker about some similar problems that describes their attitude towards the problem: https://github.com/pika/pika/issues/856

Long story short it looks like they don't intend to make this a concern of Pika and it's a responsibility of whatever uses Pika. 

Any thoughts here? At this point I am struggling with this issue in some heavy production workloads. I'd ideally like to fix the root issue but if not, I am comfortable abandoning RabbitMQ entirely and jumping to Redis if this is likely less of an issue in that environment.

**Comment by Bogdanp** [comment]
What version of Dramatiq are you using? Trying to follow the stack trace, line 206 in `rabbitmq.py` takes me to a docstring so I assume it's not the latest. I ran into the same issue and I believe I fixed it in version 0.10 (latest is 0.19) by making the `enqueue` method reconnect when it encounters a closed connection. If you are running an older version then simply upgrading will probably fix the issue.  If it doesn't, let me know and I'll take a deeper look. Thanks!

**Comment by whalesalad** [comment]
Will give that a spin, thanks for the response! I'll update this thread when I know more.

**Comment by whalesalad** [comment]
Looks like the update worked. Going to close this out. Thanks!

**Comment by Bogdanp** [comment]
Good to hear. Thanks!
